src/internal/executor/executor.cpp

```diff
@@ -91,4 +102,41 @@ void Executor::do_service_await_join()
     m_pipeline_manager->service_await_join();
 }
 
+// convert to std::expect
+bool valid_pipeline(const pipeline::Pipeline& pipeline)
+{
+    bool valid = true;
+    pipeline::PortGraph pg(pipeline);
+
+    for (const auto& [name, connections] : pg.port_map())
+    {
+        // first validate all port names have at least one:
+        // - segment using that port as an ingress, and
+        // - segment using that port as an egress
+        if (connections.egress_segments.empty() or connections.ingress_segments.empty())
+        {
+            valid = false;
+            // todo - print list of segments names for ingress/egres connections to this port
+            LOG(WARNING) << "port: " << name << " has incomplete connections - used as ingress on "
+                         << connections.ingress_segments.size() << " segments; used as egress on "
+                         << connections.egress_segments.size() << " segments";
+        }
+
+        // we currently only have an load-balancer manifold
+        // it doesn't make sense to connect segments of different types to a load-balancer, they should probably be
```

Comment:
What 'types' of segments are there? Is this just referencing the type of the ingress/egress port on the segment, or something else?

Reply:
By types, I mean two different segments with the same port name.

https://github.com/ryanolson/srf/blob/validate_ports/src/tests/test_pipeline.cpp#L324-L325

When you have two different types of segments, it's unclear that load-balancing between them makes sense. We definitely need more information to handle this condition properly.